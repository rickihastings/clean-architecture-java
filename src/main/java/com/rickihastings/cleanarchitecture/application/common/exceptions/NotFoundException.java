package com.rickihastings.cleanarchitecture.application.common.exceptions;

public class NotFoundException extends Exception {
    public NotFoundException(String error) {
        super(error);
    }
}
