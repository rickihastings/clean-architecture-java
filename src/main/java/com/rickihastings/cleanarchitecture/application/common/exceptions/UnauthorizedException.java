package com.rickihastings.cleanarchitecture.application.common.exceptions;

import an.awesome.pipelinr.AggregateException;

import java.util.List;

public class UnauthorizedException extends AggregateException {
    public UnauthorizedException() {
        super(List.of(new Exception("Unauthorized")));
    }
}
