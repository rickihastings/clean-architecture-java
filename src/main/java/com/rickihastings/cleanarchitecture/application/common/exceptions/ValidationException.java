package com.rickihastings.cleanarchitecture.application.common.exceptions;

import an.awesome.pipelinr.AggregateException;

import java.util.Collection;

public class ValidationException extends AggregateException {
    public ValidationException(Collection<Throwable> exceptions) {
        super(exceptions);
    }
}
