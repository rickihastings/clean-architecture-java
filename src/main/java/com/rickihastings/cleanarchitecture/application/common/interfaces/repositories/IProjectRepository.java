package com.rickihastings.cleanarchitecture.application.common.interfaces.repositories;

import com.rickihastings.cleanarchitecture.domain.entities.Project;
import org.springframework.stereotype.Repository;
import java.util.List;
import java.util.Optional;

@Repository
public interface IProjectRepository {
    List<Project> findAll();
    Optional<Project> findById(Long id);
    Project save(Project project);
}
