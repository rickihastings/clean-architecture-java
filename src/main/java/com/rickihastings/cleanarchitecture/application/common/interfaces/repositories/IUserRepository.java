package com.rickihastings.cleanarchitecture.application.common.interfaces.repositories;

import com.rickihastings.cleanarchitecture.domain.User;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface IUserRepository {
    Optional<User> getUserByUsername(String username);
    User save(User save);
}
