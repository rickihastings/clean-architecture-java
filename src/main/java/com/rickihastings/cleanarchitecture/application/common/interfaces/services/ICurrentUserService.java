package com.rickihastings.cleanarchitecture.application.common.interfaces.services;

import com.rickihastings.cleanarchitecture.domain.User;
import org.springframework.stereotype.Service;

@Service
public interface ICurrentUserService {
    User getUser();
}
