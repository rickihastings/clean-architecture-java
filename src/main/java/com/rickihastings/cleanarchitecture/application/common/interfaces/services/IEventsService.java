package com.rickihastings.cleanarchitecture.application.common.interfaces.services;

import an.awesome.pipelinr.Notification;

public interface IEventsService {
    <E extends Notification> void publish(E event);
}
