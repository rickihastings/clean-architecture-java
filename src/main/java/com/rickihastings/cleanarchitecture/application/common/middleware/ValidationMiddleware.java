package com.rickihastings.cleanarchitecture.application.common.middleware;

import an.awesome.pipelinr.Command;
import com.rickihastings.cleanarchitecture.application.common.exceptions.ValidationException;
import org.springframework.stereotype.Component;
import javax.validation.ConstraintViolation;
import javax.validation.Validator;
import javax.validation.ValidatorFactory;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Set;

@Component
public class ValidationMiddleware implements Command.Middleware {

    private final Validator validator;

    public ValidationMiddleware(ValidatorFactory factory) {
        this.validator = factory.getValidator();
    }

    @Override
    public <R, C extends Command<R>> R invoke(C command, Next<R> next) {
        Set<ConstraintViolation<C>> violations = validator.validate(command);

        if (violations.size() > 0) {
            Collection<Throwable> exceptions = new ArrayList<>();
            for (ConstraintViolation<C> violation : violations) {
                exceptions.add(new Exception(violation.getMessage()));
            }

            throw new ValidationException(exceptions);
        }

        return next.invoke();
    }
}
