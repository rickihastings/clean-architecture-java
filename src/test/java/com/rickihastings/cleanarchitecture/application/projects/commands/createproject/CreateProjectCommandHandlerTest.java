package com.rickihastings.cleanarchitecture.application.projects.commands.createproject;

import an.awesome.pipelinr.Pipelinr;
import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.rickihastings.cleanarchitecture.DefaultSnapshotSerializer;
import com.rickihastings.cleanarchitecture.application.common.exceptions.ValidationException;
import com.rickihastings.cleanarchitecture.application.common.interfaces.repositories.IProjectRepository;
import com.rickihastings.cleanarchitecture.application.common.interfaces.services.ICurrentUserService;
import com.rickihastings.cleanarchitecture.application.common.middleware.ValidationMiddleware;
import com.rickihastings.cleanarchitecture.seeds.UserSeeds;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mockito;

import javax.validation.Validation;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.when;

@ExtendWith({ SnapshotExtension.class })
public class CreateProjectCommandHandlerTest {

    private Expect expect;

    private final ICurrentUserService currentUserService = Mockito.mock(ICurrentUserService.class);
    private final IProjectRepository projectRepository = Mockito.mock(IProjectRepository.class);

    private Pipelinr pipeline;

    @BeforeEach
    void setup() {
        var factory = Validation.buildDefaultValidatorFactory();

        pipeline = new Pipelinr()
                .with(() -> Stream.of(new CreateProjectCommandHandler(currentUserService, projectRepository)))
                .with(() -> Stream.of(new ValidationMiddleware(factory)));
    }

    @Test
    public void shouldCreateProjectWhenValid() {
        var createProjectCommand = new CreateProjectCommand();
        createProjectCommand.setTitle("Test Project");
        when(currentUserService.getUser()).thenReturn(UserSeeds.getUser("joebloggs").get());

        var result = pipeline.send(createProjectCommand);

        expect.serializer(DefaultSnapshotSerializer.class).toMatchSnapshot(result);
    }

    @Test
    public void shouldThrowValidationErrorTitleNotNull() {
        assertThrows(ValidationException.class, () -> pipeline.send(new CreateProjectCommand()));
    }

    @Test
    public void shouldThrowValidationErrorTitleNotBlank() {
        assertThrows(ValidationException.class, () -> {
            var createProjectCommand = new CreateProjectCommand();
            createProjectCommand.setTitle("");

            pipeline.send(createProjectCommand);
        });
    }
}
