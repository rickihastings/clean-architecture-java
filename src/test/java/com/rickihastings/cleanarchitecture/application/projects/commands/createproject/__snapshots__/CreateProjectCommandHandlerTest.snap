com.rickihastings.cleanarchitecture.application.projects.commands.createproject.CreateProjectCommandHandlerTest.shouldCreateProjectWhenValid=[
  {
    "title": "Test Project"
  }
]